### DISCUSSION TITLE
Choosing default SYCL devices for easy switching between CPU and GPU?

### DISCUSSION THREAD

**Discussion by Kiguli** [AdaptiveCpp/AdaptiveCpp#1183] [Q&A] [answered]
Hi,

I have been using AdaptiveCpp for a tool I am developing to help with large-scale computations and I want to give the users the opportunity to choose whether they run the code on CPU or GPU using the default SYCL options. I am struggling to find info on how I can set this up, such as in a configuration file in order to keep the messy implementation details in the background so the user doesn't need to see them. 

Does someone have some helpful links they can send me or a little example of how I could let the user choose for themselves which implementation they wish to use?

Thanks for any advice in advance :)

**Comment by illuhad** [comment]
What do you mean by `using the default SYCL options`?

You have two options:
* Either set the environment variable `ACPP_VISIBILITY_MASK` such that only the CPU backend is exposed when GPUs should not be used: `export ACPP_VISIBILITY_MASK=omp`
* Programmatically, change your program such that it constructs a `queue` for either CPU or GPU based on user input, e.g. using `sycl::queue q{sycl::cpu_selector_v}` or `sycl::queue q{sycl::gpu_selector_v}`, or just query the available devices using `sycl::device::get_devices()` and construct the queue with the chosen device.

**Reply by Kiguli** [reply]
Hi, thanks for your reply.

I am wondering if I can set the choice between two options as a variable in my main file so that the user can choose between CPU and GPU without having to either edit the sycl::queue q{type} themselves or set any environment variables. I guess I could implement two functions one with cpu_selector and one with gpu_selector but that would be quite cumbersome. The user will have a main file they can adjust the parameters inside of, before compiling and running, but the actual functions will be hidden.

**Reply by illuhad** [reply, marked as answer]
Something like this?
```c++

sycl::qeue get_queue(bool use_gpu) {
  if(use_gpu) {
    return sycl::queue{sycl::gpu_selector_v};
  else
    return sycl::queue{sycl::cpu_selector_v};
}

// Uses the queue somehow
void run(sycl::queue& q);

int main() {
  sycl::queue q = get_queue(/* obtain parameter of whether to use GPU from somewhere*/);
  run(q);
}

```

**Reply by Kiguli** [reply]
Thank you, I think that's what I will do. So it can't be set globally for SYCL to work out itself, unless using the environment variables. Thank you so much :)
